File: src/main/julia/AoC2017_13.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2017_13

include("aoc.jl")

struct Layer
    depth::Int
    range::Int
end

function Layer((a, b)::Tuple)
    return Layer(a, b)
end

function _parse(input)
    return [Layer(Tuple(parse.(Int, split(line, ": ")))) for line ∈ input]
end

function caught(layer::Layer, delay::Int)
    return (layer.depth + delay) % ((layer.range - 1) * 2) == 0
end

function part1(input)
    layers = _parse(input)
    return sum(layer -> layer.depth * layer.range,
               (layer for layer ∈ layers if caught(layer, 0)))
end

function part2(input)
    layers = _parse(input)
    delay::Int = 0
    while any(caught(layer, delay) for layer ∈ layers)
        delay += 1
    end
    return delay
end

TEST = @aoc_splitlines("""
0: 3
1: 2
4: 4
6: 4
""")

function samples()
    @assert part1(TEST) == 24
    @assert part2(TEST) == 10
end

end # module AoC2017_13

aoc_main(@__FILE__, ARGS, 2017, 13)

File: src/main/python/AoC2017_13.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 13
#

from aoc import my_aocd
from typing import NamedTuple
import aocd


class Layer(NamedTuple):
    depth: int
    range: int


def _parse(inputs: tuple[str]) -> list[Layer]:
    return [Layer(*map(int, layer.split(": "))) for layer in inputs]


def _caught(layer: Layer, delay: int) -> bool:
    return (layer.depth + delay) % ((layer.range - 1) * 2) == 0


def part_1(inputs: tuple[str]) -> int:
    layers = _parse(inputs)
    return sum(map(lambda layer: layer.depth * layer.range,
                   (layer for layer in layers if _caught(layer, 0))))


def part_2(inputs: tuple[str]) -> int:
    layers = _parse(inputs)
    delay = 0
    while any(_caught(layer, delay) for layer in layers):
        delay += 1
    return delay


TEST = '''\
0: 3
1: 2
4: 4
6: 4
'''.splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 13)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 24
    assert part_2(TEST) == 10

    inputs = my_aocd.get_input(2017, 13, 43)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()
